### ISSUE TITLE
GOES FITS files 1980-2014 which cause GOESLightCurve to crash

### ISSUE THREAD

**Issue by DanRyanIrish** [sunpy/sunpy#1337]
Here are a list of dates in the period 1980-2014 for which the corresponding GOES FITS files cause the GOESLightCurve object to crash when created.  The files are listed by the reasons they crash/errors that are raised.  These dates were discovered as part of PR #1319.

File exists but no data inside
1995-04-23
1995-04-24
1995-04-25
1995-04-26
1995-04-27
1995-04-28
1995-04-29
1995-04-30
1995-05-01
1995-05-02
1995-05-03
1995-05-04
1995-05-05
1995-05-06
1995-05-07
1995-05-08
1995-05-09
1995-05-10
1995-05-11
1995-05-12
1995-05-13
1995-05-14
1995-05-15
1995-05-16
1995-05-17
1995-05-18

File Doesn't Exist
1980-01-27
1980-01-28
1980-01-29
1980-01-30
1980-01-31
1980-02-01
1980-02-02
1980-02-03
1980-02-04
1980-02-17
1980-02-18
1980-02-19
1980-02-20
1980-02-21
1980-02-22
1980-02-23
1980-05-08
1980-05-09
1980-05-10
1980-05-11
1980-11-01
1981-01-27
1981-01-28
1981-01-29
1981-01-30
1981-01-31
1981-02-04
1981-02-05
1981-02-06
1981-02-17
1981-02-18
1981-02-19
1981-09-06
1981-09-07
1981-11-04
1981-11-05
1981-11-06
1981-11-07
1982-01-01
1982-01-01
1982-02-03
1982-02-04
1982-02-05
1982-02-06
1982-10-31
1983-02-13
1983-05-18
1983-06-01
1983-06-02
1983-06-03
1983-06-04
1983-06-05
1983-06-06
1983-06-07
1983-06-08
1983-06-09
1983-06-10
1983-06-11
1983-06-12
1983-06-13
1983-06-14
1983-06-15
1983-06-16
1983-06-17
1983-06-18
1983-06-19
1983-06-20
1983-06-21
1983-06-22
1983-06-23
1983-06-24
1983-06-25
1983-06-26
1983-06-27
1983-06-28
1983-06-29
1983-06-30
1983-09-06
1983-09-19
1983-11-05
1984-05-27
1984-07-30
1984-07-31
1984-08-21
1984-08-22
1984-08-23
1984-08-24
1984-08-25
1984-08-26
1984-12-16
1984-12-21
1984-12-22
1984-12-23
1984-12-24
1984-12-25
1984-12-26
1984-12-27
1984-12-28
1984-12-29
1984-12-30
1984-12-31
1985-01-01
1985-01-01
1985-06-30
1986-09-30
1987-01-07
1987-01-08
1987-01-09
1987-01-10
1987-01-11
1987-01-12
1987-01-13
1987-01-14
1987-01-15
1987-01-16
1987-01-17
1987-01-18
1987-01-19
1987-01-20
1987-01-21
1987-01-22
1987-01-23
1987-01-24
1987-01-25
1987-01-26
1987-01-27
1987-01-28
1987-01-29
1987-01-30
1987-04-07
1987-04-08
1989-07-01
1992-08-15
1992-12-09
1993-04-04
1993-07-24
1993-11-23
1993-11-24
1993-11-25
1993-11-26
1993-11-27
1993-11-28
1993-11-29
1993-11-30
1993-12-12
1994-07-09
1994-07-10
1994-07-31
1995-10-03
1995-10-07
1995-10-08
1995-12-28
1996-01-07
1996-04-28
1996-08-07
1996-08-08
1996-10-03
1997-05-24
1997-10-18
1997-10-19
1998-03-21
1999-01-17
1999-01-18
1999-01-19
1999-01-21
2000-02-03
2000-02-04
2000-02-05
2000-02-06
2000-06-07
2001-03-25
2002-06-25
2002-06-26
2002-06-27
2002-06-28
2002-06-29
2002-06-30
2002-07-01
2003-04-16
2003-04-25
2003-04-26
2003-04-27
2003-04-28
2003-04-29
2003-04-30
2003-05-01
2003-05-02
2003-05-03
2003-05-04
2003-07-23
2003-07-24
2003-07-25
2003-07-26
2003-07-27
2003-07-28
2003-09-19
2004-02-09
2004-02-26
2004-02-27
2004-02-28
2004-02-29
2004-04-13
2004-09-17
2005-11-16
2005-11-21
2005-11-27
2006-06-20
2006-06-21
2006-06-22
2006-06-23
2006-06-24
2006-06-25
2006-06-26
2006-06-27
2006-06-28
2006-06-29
2006-06-30
2006-07-01
2006-07-02
2006-07-03
2006-07-04
2006-07-05
2006-07-06
2006-07-07
2006-07-08
2006-07-09
2006-07-10
2006-07-11
2006-07-12
2006-07-13
2006-07-14
2006-07-15
2006-07-16
2006-07-17
2006-07-18
2006-07-19
2006-07-20
2006-07-21
2006-07-22
2006-07-23
2006-07-24
2006-07-25
2006-07-26
2006-07-27
2006-07-28
2006-07-29
2006-07-30
2006-07-31
2006-08-01
2006-08-02
2006-08-03
2006-08-04
2006-08-05
2006-08-06
2006-08-07
2006-08-08
2006-08-09
2006-08-10
2006-08-11
2006-08-12
2006-08-13
2006-08-14
2006-08-15
2006-08-16
2006-08-17
2006-08-18
2006-08-19
2006-08-20
2006-08-21
2006-08-22
2006-08-23
2006-08-24
2006-08-25
2006-08-26
2006-08-27
2006-08-28
2006-08-29
2006-08-30
2006-08-31
2006-09-01
2006-09-02
2006-09-03
2006-09-04
2006-09-05
2006-09-06
2006-09-07
2006-09-08
2006-09-09
2006-09-10
2006-09-11
2006-09-12
2006-09-13
2006-09-14
2006-09-15
2006-09-16
2006-09-17
2006-09-18
2006-09-19
2006-09-20
2006-09-21
2006-09-22
2006-09-23
2006-09-24
2006-09-25
2006-09-26
2006-09-27
2006-09-28
2006-09-29
2006-09-30
2006-10-01
2006-10-02
2006-10-03
2006-10-04
2006-10-05
2006-10-06
2006-10-07
2006-10-08
2006-10-09
2006-10-10
2006-10-11
2006-10-12
2006-10-13
2006-10-14
2006-10-15
2006-10-16
2006-10-17
2006-10-18
2006-10-19
2006-10-20
2006-10-21
2006-10-22
2006-10-23
2006-10-24
2006-10-25
2006-10-26
2006-10-27
2006-10-28
2006-10-29
2006-10-30
2006-10-31
2006-11-01
2006-11-02
2006-11-03
2006-11-04
2006-11-05
2006-11-06
2006-11-07
2006-11-08
2006-11-09
2006-11-10
2006-11-11
2006-11-12
2006-11-13
2006-11-14
2006-11-15
2006-11-16
2006-11-17
2006-11-18
2006-11-19
2006-11-20
2006-11-21
2006-11-22
2006-11-23
2006-11-24
2006-11-25
2006-11-26
2006-11-27
2006-11-28
2006-11-29
2006-11-30
2006-12-01
2006-12-02
2006-12-03
2006-12-04
2006-12-05
2006-12-06
2006-12-07
2006-12-08
2006-12-09
2006-12-10
2006-12-11
2006-12-12
2006-12-13
2006-12-14
2006-12-15
2006-12-16
2006-12-17
2006-12-18
2006-12-19
2006-12-20
2006-12-21
2006-12-22
2006-12-23
2006-12-24
2006-12-25
2006-12-26
2006-12-27
2006-12-28
2006-12-29
2006-12-30
2006-12-31
2007-01-01
2007-01-01
2007-01-02
2007-01-03
2007-01-04
2007-01-05
2007-01-06
2007-01-07
2007-01-08
2007-01-09
2007-01-10
2007-01-11
2007-01-12
2007-01-13
2007-01-14
2007-01-15
2007-01-16
2007-01-17
2007-01-18
2007-01-19
2007-01-20
2007-01-21
2007-01-22
2007-01-23
2007-01-24
2007-01-25
2007-01-26
2007-01-27
2007-01-28
2007-01-29
2007-01-30
2007-01-31
2007-02-01
2007-02-02
2007-02-03
2007-02-04
2007-02-05
2007-02-06
2007-02-07
2007-02-08
2007-02-09
2007-02-10
2007-02-11
2007-02-12
2007-02-13
2007-02-14
2007-02-15
2007-02-16
2007-02-17
2007-02-18
2007-02-19
2007-02-20
2007-02-21
2007-02-22
2007-02-23
2007-02-24
2007-02-25
2007-02-26
2007-02-27
2007-02-28
2007-03-01
2007-03-02
2007-03-03
2007-03-04
2007-03-05
2007-03-06
2007-03-07
2007-03-08
2007-03-09
2007-03-10
2007-03-11
2007-03-12
2007-03-13
2007-03-14
2007-03-15
2007-03-16
2007-03-17
2007-03-18
2007-03-19
2007-03-20
2007-03-21
2007-03-22
2007-03-23
2007-03-24
2007-03-25
2007-03-26
2007-03-27
2007-03-28
2007-03-29
2007-03-30
2007-03-31
2007-04-01
2007-04-02
2007-04-03
2007-04-04
2007-04-05
2007-04-06
2007-04-07
2007-04-08
2007-04-09
2007-04-10
2007-04-11
2007-04-12
2007-04-13
2007-04-14
2007-04-15
2007-04-16
2007-04-17
2007-04-18
2007-04-19
2007-04-20
2007-04-21
2007-04-22
2007-04-23
2007-04-24
2007-04-25
2007-04-26
2007-04-27
2007-04-28
2007-04-29
2007-04-30
2007-05-01
2007-05-02
2007-05-03
2007-05-04
2007-05-05
2007-05-06
2007-05-07
2007-11-21
2007-11-22
2007-11-23
2007-11-24
2007-11-25
2007-11-26
2007-11-27
2007-11-28
2007-11-29
2007-11-30
2007-12-01
2007-12-02
2007-12-03
2007-12-04
2007-12-05
2007-12-06
2007-12-07
2007-12-08
2007-12-09
2007-12-10
2007-12-11
2007-12-12
2007-12-13
2007-12-14
2007-12-15
2007-12-16
2007-12-17
2007-12-18
2007-12-19
2007-12-20
2007-12-21
2007-12-22
2007-12-23
2007-12-24
2007-12-25
2007-12-26
2007-12-27
2007-12-28
2007-12-29
2007-12-30
2007-12-31
2008-01-01
2008-01-01
2008-01-02
2008-01-03
2008-01-04
2008-01-05
2008-01-06
2008-01-07
2008-01-08
2008-01-09
2008-01-10
2008-01-11
2008-01-12
2008-01-13
2008-01-14
2008-01-15
2008-01-16
2008-01-17
2008-01-18
2008-01-19
2008-01-20
2008-01-21
2008-01-22
2008-01-23
2008-01-24
2008-01-25
2008-01-26
2008-01-27
2008-01-28
2008-01-29
2008-01-30
2008-01-31
2008-02-01
2008-02-02
2008-02-03
2008-02-04
2008-02-05
2008-02-06
2008-02-07
2008-02-08
2008-02-09
2008-02-10
2012-03-22
2012-10-23
2012-10-24
2012-10-25
2012-11-01
2012-11-02
2012-11-03
2012-11-04
2012-11-08
2012-11-14
2012-11-15
2014-12-05

File Doesn't Exist (but a No Satellite Error is raised due to basing Satellite periods off which existing FITS files were created with data from each satellite)
1983-04-29
1983-04-30
1983-05-01
1983-05-02

KeyError
1981-01-08
1982-05-21
1982-08-14
1983-05-26
1983-12-25
1984-01-16
1984-08-14
1984-10-09
1985-03-17
1985-04-22
1985-07-29
1995-12-27
2000-01-18
2010-03-14

**Comment by Cadair** [comment]
yay! The `KeyError` ones imply we are not reading the layout properly? or are they corrupt in some way?

**Comment by DanRyanIrish** [comment]
Not sure at the moment.  Haven't had a chance to figure that out yet.

**Comment by Rishabhmishra328** [comment]
Hi,
Does the issue involve invoking an error when data from these dates are called?

**Comment by DanRyanIrish** [comment]
@Rishabhmishra328 If you try to create a GOESLightCurve object for these dates, you do indeed get an error.

**Comment by Rishabhmishra328** [comment]
@DanRyanIrish ,
Let me rephrase my question,
To solve this issue do we need to produce an error message when a user calls for GOESLightCurve objest for the following dates?

**Comment by DanRyanIrish** [comment]
@Rishabhmishra328 I think that's a philosophical question that we should discuss further.  @Cadair @ehsteve @aringlis @ayshih et al., any thoughts?

**Comment by Cadair** [comment]
I presume the error at the moment is something along the lines of "File not found"?

That makes sense to me, at the end of the day, the user should realise we are talking about looking at the server.

**Comment by DanRyanIrish** [comment]
@Cadair.  Of course the error depends.  Some files exist but are empty, some files exist and lead to a KeyError, some files do not exist.  So it depends.

**Comment by Rishabhmishra328** [comment]
@DanRyanIrish @Cadair 
But whatever it is, if the user is not receiving any data, he/she should know that there is a fault in the data acquiring object and not his/her code.
I believe it would be best if we can add a error message on the dates which have data issues (Empty data files, KeyError, FIle doesn't exist etc.).

**Comment by aringlis** [comment]
The KeyError happens because, for some reason, the time values from these specific GOES files are not completely in order. This leads to a pandas truncate error. For example:

    from astropy.io import fits
    import pandas
    from numpy import floor
    from sunpy.time import parse_time
    import numpy as np

    hdu = fits.open('go06850317.fits')
    h1 = hdu[0]
    xrsb = h1.data[2,:]
    xrsa = h1.data[1,:]
    start_time = parse_time(h1.header['TIMEZERO'])
    times = [start_time + datetime.timedelta(seconds=int(floor(s)),
            microseconds=int((s - floor(s)) * 1e6)) for s in seconds_from_start]
    dat = pandas.DataFrame({'xrsa': newxrsa, 'xrsb': newxrsb},index = times)

and then:

    test = dat.truncate(parse_time('1985-03-17 02:00'),parse_time('1985-03-17 04:00'))
    ERROR: KeyError: 479872800000000000 [IPython.core.interactiveshell]

You can see that the time indices are out of order:

    difftimes = times[1:-1] - times[0:-2]  
    np.min(difftimes) 
    Out[76]: -6.0    

What happens is that the indices in the pandas dataframe need to be in sequential order for the pandas truncate function to work, otherwise it throws this (pretty unhelpful) KeyError. This can be worked around by sorting the Dataframe first, e.g.

    dat.sort(inplace=True)
    test = dat.truncate(parse_time('1985-03-17 02:00'),parse_time('1985-03-17 04:00'))

This problem is familiar - I remember we encountered it somewhere before, but i don't remember where (@wafels, @Cadair?). Anyhow, this is really a separate bug issue - there is nothing wrong with the GOES data in these files, so far as I can tell. If one doesn't exist, I'll raise a separate issue about this - in principle it could happen for any lightcurve data file from any instrument.

**Comment by DanRyanIrish** [comment]
Ah!!  So that's why!  Well worked out @aringlis

**Comment by Rishabhmishra328** [comment]
@aringlis, this is good. 
This might solve the issue of KeyError but the empty data file error will still startle the user to what went wrong.

**Comment by Rishabhmishra328** [comment]
@DanRyanIrish @Cadair 
Can you please guide me how to solve this issue.
Does it need an error message on data request of these dates or anything else?

**Comment by wtbarnes** [comment]
Does this still persist in `timeseries`?

**Comment by dstansby** [comment]
Since the error message doesn't seem to be posted above, and I can't easily find one of the offending files, I'm going to close. If anyone can provide the error message and a file that causes it, feel free to re-open.
